### PR TITLE
loading-button

### PR DESCRIPTION
**Pull request by Vineet119**

### CODE REVIEW

File: site/docs/components/button/examples.mdx [hunk 1]
```diff
@@ -79,5 +79,34 @@ A button with the prop `disabled={true}` will suppress all functionality. If you
 
 Use full width buttons on mobile devices or smaller viewport, the button will take up the full width of its container.
 
+</LivePreview>
+
+<LivePreview componentName="button" exampleName="LoadingButtons" >
+
+## Loading Buttons
+
+The buttons will show a loading spinner when setting the isLoading prop to true. You should pass a message using the loadingText prop for screenreader to announce while loading state is true. It defaults to "Loading".
```
Comment (bhoppers2008) [inline site/docs/components/button/examples.mdx:88]: Prop names and values should be marked up i believe: `loadingText` `true` etc

File: packages/core/src/__tests__/__e2e__/button/Button.cy.tsx [hunk 2]
```diff
@@ -6,58 +6,70 @@ const composedStories = composeStories(buttonStories);
 const { FeatureButton, FocusableWhenDisabled } = composedStories;
 
 describe("Given a Button", () => {
-  checkAccessibility(composedStories);
+  // checkAccessibility(composedStories);
```
Comment (origami-z) [inline packages/core/src/__tests__/__e2e__/button/Button.cy.tsx:9]: remove comments

File: packages/core/src/button/Button.css [hunk 3]
```diff
@@ -91,23 +91,23 @@
...
+.saltButton.saltButton-active:focus-visible:not(.saltButton-loading-primary, .saltButton-loading-secondary, .saltButton-loading-cta),
+.saltButton:focus-visible(:active):not(.saltButton-loading-primary, .saltButton-loading-secondary, .saltButton-loading-cta) {
   background: var(--saltButton-background-active-hover, var(--button-background));
   color: var(--saltButton-text-color-active-hover, var(--button-text-color));
   border-color: var(--saltButton-borderColor-hover, var(--button-borderColor-hover));
 }
 
 /* Pseudo-class applied to the root element on hover when Button is not active or disabled */
-.saltButton:hover {
+.saltButton:hover:not(.saltButton-loading-primary, .saltButton-loading-secondary, .saltButton-loading-cta) {
```
Comment (origami-z) [inline packages/core/src/button/Button.css:102]: Do we really need these increased specificity? Can we do it another way by using `--button-x` variables?

File: packages/core/src/button/Button.tsx [hunk 4]
```diff
@@ -26,6 +27,19 @@ export interface ButtonProps extends ComponentPropsWithoutRef<"button"> {
...
   variant?: ButtonVariant;
+  /**
+   * To show a loading spinner.
+   */
+  isLoading?: boolean;
+  /**
+   * For the screen reader to announce while button is in loading state
+   * 'Loading' is the default value.
+   */
+  loadingText?: string;
```
Comment (origami-z) [inline packages/core/src/button/Button.tsx:38]: Do we really need a new prop to deal with announcement? 
cc @joshwooding

File: packages/core/src/button/Button.css (hunk 3 above)
Comment (Vineet119) [reply packages/core/src/button/Button.css]: Done

File: packages/core/src/button/Button.tsx (hunk 4 above)
Comment (joshwooding) [reply packages/core/src/button/Button.tsx:38]: Without clear accessibility guidance I don't want any of these added apart from the isLoading prop

File: packages/core/src/button/Button.tsx [hunk 5]
```diff
@@ -26,6 +27,19 @@ export interface ButtonProps extends ComponentPropsWithoutRef<"button"> {
    * 'primary' is the default value.
    */
   variant?: ButtonVariant;
+  /**
+   * To show a loading spinner.
+   */
+  isLoading?: boolean;
```
Comment (joshwooding) [inline packages/core/src/button/Button.tsx:33]: This does not follow our standard prop naming, "is" shouldn't be the prefix

File: packages/core/src/button/Button.tsx [hunk 6]
```diff
@@ -72,15 +89,47 @@ export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
...
       >
-        {children}
+        {isLoading ? (
+          <>
+            <span className={clsx(withBaseName("loading-overlay"), className)}>
+              <Spinner
+                size="small"
+                className={clsx(withBaseName("loading-spinner"), className)}
+              />
+              <span
```
Comment (joshwooding) [inline packages/core/src/button/Button.tsx:110]: Even if this is the right handling of accessibility, why is this used over the built-in controls for Spinner?

File: packages/core/stories/button/button.stories.tsx [hunk 7]
```diff
@@ -168,3 +172,111 @@ export const FullWidth: StoryFn<typeof Button> = () => {
...
+        {secondaryButtonLabel}
+        {withRightArrowIcon && <ChevronRightIcon aria-hidden />}
+      </Button>
+    </div>
+  );
+};
+
+export const LoadingButtons: StoryFn<typeof Button> = () => {
+  return (
+    <LoadingButtonGrid
```
Comment (joshwooding) [inline packages/core/stories/button/button.stories.tsx:262]: I'm not a big fan of these examples. This looks like it's similar to a template but it's not using the storybook method

Comment (joshwooding) [reply packages/core/stories/button/button.stories.tsx:262]: I'm also not a fan of changing labels via props either

Comment (Vineet119) [reply packages/core/stories/button/button.stories.tsx]: done

File: packages/core/src/button/Button.tsx (hunk 5 above)
Comment (Vineet119) [reply packages/core/src/button/Button.tsx]: removed

File: docs/css/salt-countries.css [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/* Style applied to the root element */
```
Comment (joshwooding) [inline docs/css/salt-countries.css:1]: This shouldn't be committed

File: packages/core/src/button/Button.tsx [hunk 9]
```diff
@@ -72,15 +89,41 @@ export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
           withBaseName(variant),
           {
             [withBaseName("disabled")]: disabled,
+            [withBaseName(`loading-${variant}`)]: loading,
             [withBaseName("active")]: active,
           },
           className
         )}
+        aria-disabled={disabled}
```
Comment (joshwooding) [inline packages/core/src/button/Button.tsx:97]: aria-disabled is handled in useButton, this overrides it?

File: packages/core/src/button/Button.tsx [hunk 10]
```diff
@@ -72,15 +89,41 @@ export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
           withBaseName(variant),
           {
             [withBaseName("disabled")]: disabled,
+            [withBaseName(`loading-${variant}`)]: loading,
```
Comment (joshwooding) [inline packages/core/src/button/Button.tsx:92]: We already have a variant class, why do we need to add loading-(variant) classes?

Comment (Vineet119) [reply packages/core/src/button/Button.tsx:92]: we need it, for handling css part, of different variant, like border for different variants, while loiading

File: packages/core/src/button/Button.tsx (hunk 9 above)
Comment (Vineet119) [reply packages/core/src/button/Button.tsx]: removed

File: docs/css/salt-countries.css (hunk 8 above)
Comment (Vineet119) [reply docs/css/salt-countries.css]: removed

File: packages/core/src/button/Button.tsx (hunk 4 above)
Comment (mark-tate) [reply packages/core/src/button/Button.tsx:38]: Now we see this in code ...Should this move to it's own component or are we comfortable having one API inside another ? On other component library's they just have one prop which drives the state.

File: packages/core/src/button/Button.tsx [hunk 11]
```diff
@@ -72,15 +89,40 @@ export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
...
             [withBaseName("active")]: active,
           },
           className
         )}
+        aria-live={loading !== undefined ? "assertive" : undefined}
```
Comment (joshwooding) [inline packages/core/src/button/Button.tsx:97]: I can't remember the outcome but what was decided on the usage of assertive here?

File: packages/core/stories/button/button.stories.tsx [hunk 12]
```diff
@@ -1,9 +1,29 @@
...
+  useId,
+  Dialog,
+  DialogHeader,
+  DialogActions,
+  DialogContent,
+  FormField,
+  FormFieldLabel,
+  Input,
+} from "@salt-ds/core";
+import { ChevronRightIcon, DoubleChevronRightIcon } from "@salt-ds/icons";
```
Comment (joshwooding) [inline packages/core/stories/button/button.stories.tsx:19]: This important should be merged in the one below.

File: packages/core/stories/button/button.stories.tsx [hunk 13]
```diff
@@ -168,3 +188,232 @@ export const FullWidth: StoryFn<typeof Button> = () => {
...
+              This page has updates. Refresh page if you want to see recent
+              changes.
+            </Text>
+            <FlowLayout gap={1} justify="end">
+              <Button
+                variant="cta"
+                loading={loadingState}
+                showLoadingText
+                loadingText="Refreshing"
+                aria-describedby=""
```
Comment (joshwooding) [inline packages/core/stories/button/button.stories.tsx:347]: Is this needed?

File: packages/core/stories/button/button.stories.tsx (hunk 12 above)
Comment (Vineet119) [reply packages/core/stories/button/button.stories.tsx]: done

File: packages/core/stories/button/button.stories.tsx (hunk 13 above)
Comment (Vineet119) [reply packages/core/stories/button/button.stories.tsx]: removed

**Comment by joshwooding** [comment]
Succeded by #3887

**Pull request closed without merging**